### PR TITLE
exec.d preinstall support for charms.reactive

### PR DESCRIPTION
**Pull request by stub42**
The exec.d preinstall hooks predate me, yet here I am dragging them into the new framework.

This is a mechanism used to customize provisioned machines so the charms can actually work. Your apt-get install isn't going to work until your bonded network has been configured and connected to the right vlan, the beta firmware loaded and injected with ipsec keys so it can pull packages from the proxy discovered by avahi. At which point the charms.reactive framework can be imported and kicked off.

I'm not sure where the exec.d/$namespace/charm-pre-install tree structure comes from, but it is the standard people are using so I'm sticking with it. I see a lot of mojo specs unpacking site customization stuff into exec.d, so changing it would cause lots of unwanted fallout.

The bulk of this is lifted directly from charm-helpers. I customized it a bit, allowing it to set the unit to blocked when things fail and coping with the install hook no longer necessarily being the first hook.

### CODE REVIEW

**Comment by cmars** [comment]
As someone who's developing layered charms for deployments where this convention is used, this is very much appreciated!

**Comment by cmars** [comment]
If it doesn't land here, it should definitely go into a basenode layer.

**Comment by johnsca** [comment]
@cmars If you give this a +1 review, I'm good with merging this.  Can you test it with one of the layers you're working on?

**Comment by stub42** [comment]
I did consider adding this as a layer. However, this would still involve adding a hook into base layer to invoke the execd layer's bootstrap code. Which might be useful, but I came up blank for other use cases that need to be bootstrapped this early in the chain. I also like how all charms get it, rather than just the ones where the charm author remembered to include the layer, since it is behaviour that operators need to activate in peculiar situations on site and with deadlines to meet.

**Comment by cmars** [comment]
@johnsca Sure thing, I'll test this out tomorrow.

**Comment by cmars** [comment]
:+1: No QA though.

File: lib/charms/layer/basic.py [hunk 1]
```diff
@@ -4,11 +4,19 @@
...
 from subprocess import check_call
 
+from charms.layers.execd import execd_preinstall
+
 
 def bootstrap_charm_deps():
     """
     Set up the base charm dependencies so that the reactive system can run.
     """
+    # execd must happen first, before any attempt to install packages or
```
Comment (chuckbutler) [inline lib/charms/layer/basic.py]: I'm late to the party, but you just unlocked what exec.d / basenode really is used for. Thanks for the breakdown stub!

**Comment by stub42** [comment]
This fork is being successfully used for production deploys of the PostgreSQL charm, so is probably safe to land.

**Pull request merged**